## Import Libraries

In [5]:
import pandas as pd
import numpy as np
import os

In [6]:
#define path
path = r'C:\Users\Administrator\Desktop\DataImmersion'

In [7]:
## Importing  orders_products_merged dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path,'chapter4','Prepared data', 'ords_prods_merged.pkl'))

In [8]:
ords_prods_merge.shape


(32404859, 19)

In [9]:
 #Aggregate mean of the “order_number” column grouped by “department_id” 
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

when worked with the first 1000000 columns of ords_prods_merge dataset the department with id 16 had the highest order mean (19.46) while department with id 17  had lowest mean (11.29).

But when we checked the mean for the entire ords_prods_merge dataset, the highest order was for department id 21  and least for department id 5. Also the average order for department id 17 increased to 15.69 and the average order for department id 16 decreased to 17.66

### Finding Loyal Customers

In [10]:
#creating a loyalty flag for existing customers using the transform() and loc() functions
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

#### Loyalty Flag 

In [11]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [15]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','max','min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal customer    10.386336  99999.0  1.0
New customer      13.294670  99999.0  1.0
Regular customer  12.495717  99999.0  1.0

### Finding types of Spenders 

In [16]:
#Create a spending flag for each user based on the average price across all their orders 
ords_prods_merge['avg_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [17]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    ...  add_to_cart_order  reordered  _merge    price_range_loc  \
0   ...                  5          0    both  Mid-range product   
1   ...                  1          1    both  Mid-range product   
2   ...                 20          0    both  Mid-range product   
3   ...                 10          0    both  Mid-range product   
4   ...                 11          1    both  Mid-range product   
5   ...                  7          0    both  Mid-range product   
6   ...                  2          0    both  Mid-range product   
7   ...                  1          0    both  Mid-range product   
8   ...                  1          0    both  Mid-range product   
9   ...                  2          1    both  Mid-range product   
10  ...                  1          0    both  Mid-range product   
11  ...                  1          0    both  Mid-range product   
12  ...                  3          1    both  Mid-range product   
13  ...                  4          1    both  Mid-range product   
14  ...                  8          0    both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy           Most orders        32   
1   Regularly busy  Regularly busy           Most orders        32   
2      Busiest day    Busiest days        Average orders         5   
3   Regularly busy      Least busy           Most orders         3   
4       Least busy      Least busy           Most o

In [18]:
ords_prods_merge[['user_id', 'avg_price', 'prices']].head(60)

user_id  avg_price  prices
0       138   6.935811     5.8
1       138   6.935811     5.8
2       709   7.930208     5.8
3       764   4.972414     5.8
4       764   4.972414     5.8
5       777   6.935398     5.8
6       825   5.957576     5.8
7       910   6.680000     5.8
8      1052   7.162500     5.8
9      1052   7.162500     5.8
10     1379   5.813636     5.8
11     1480   5.081633     5.8
12     1480   5.081633     5.8
13     1480   5.081633     5.8
14     1494   6.803797     5.8
15     1494   6.803797     5.8
16     1494   6.803797     5.8
17     1540   6.607122     5.8
18     1540   6.607122     5.8
19     1540   6.607122     5.8
20     1540   6.607122     5.8
21     1540   6.607122     5.8
22     1540   6.607122     5.8
23     1540   6.607122     5.8
24     1540   6.607122     5.8
25     1540   6.607122     5.8
26     1540   6.607122     5.8
27     1540   6.607122     5.8
28     1540   6.607122     5.8
29     1540   6.607122     5.8
30     1540   6.607122     5.8
31     1540   6.607122     5.8
32     1540   6.607122     5.8
33     1540   6.607122     5.8
34     1598   7.123529     5.8
35     1598   7.123529     5.8
36     1598   7.123529     5.8
37     1647   4.853521     5.8
38     1647   4.853521     5.8
39     1647   4.853521     5.8
40     1647   4.853521     5.8
41     1647   4.853521     5.8
42     2833   7.377143     5.8
43     2850   8.054153     5.8
44     2857   8.353793     5.8
45     3010   6.452800     5.8
46     3029   7.960606     5.8
47     3395   6.322222     5.8
48     3904   7.511111     5.8
49     4006   7.441100     5.8
50     4006   7.441100     5.8
51     4006   7.441100     5.8
52     4006   7.441100     5.8
53     4006   7.441100     5.8
54     4122   7.598765     5.8
55     4141   6.033333     5.8
56     4343   7.393939     5.8
57     4553   5.715789     5.8
58     4845   7.229630     5.8
59     6866   4.807143     5.8

#### Spender Flag

In [19]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10 , 'spender_flag'] = 'High spender'

In [20]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spender_flag'] = 'Low spender'

In [21]:
ords_prods_merge['spender_flag'].value_counts(dropna = False)

Low spender     31770746
High spender      634113
Name: spender_flag, dtype: int64

### Frequent versus non-frequent customers

In [22]:
ords_prods_merge['order_frequency'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(np.median)

In [23]:
ords_prods_merge[['user_id', 'days_since_prior_order', 'order_frequency']].head(60)




user_id  days_since_prior_order  order_frequency
0       138                     3.0              8.0
1       138                    20.0              8.0
2       709                     6.0              6.0
3       764                     0.0              9.0
4       764                     9.0              9.0
5       777                    26.0             11.0
6       825                    30.0             19.0
7       910                    30.0              3.5
8      1052                    19.0              9.5
9      1052                    15.0              9.5
10     1379                     0.0             20.0
11     1480                     7.0              7.0
12     1480                     7.0              7.0
13     1480                    12.0              7.0
14     1494                     2.0              6.0
15     1494                     6.0              6.0
16     1494                    10.0              6.0
17     1540                     7.0              7.0
18     1540                     6.0              7.0
19     1540                    14.0              7.0
20     1540                     8.0              7.0
21     1540                     6.0              7.0
22     1540                    28.0              7.0
23     1540                     7.0              7.0
24     1540                     7.0              7.0
25     1540                     0.0              7.0
26     1540                     7.0              7.0
27     1540                     7.0              7.0
28     1540                     7.0              7.0
29     1540                     7.0              7.0
30     1540                     3.0              7.0
31     1540                    16.0              7.0
32     1540                     5.0              7.0
33     1540                     6.0              7.0
34     1598                     0.0              2.0
35     1598                    21.0              2.0
36     1598                     2.0              2.0
37     1647                     0.0             25.0
38     1647                    18.0             25.0
39     1647                     8.0             25.0
40     1647                    30.0             25.0
41     1647                    25.0             25.0
42     2833                     0.0              5.0
43     2850                     9.0              7.0
44     2857                    10.0             10.0
45     3010                     7.0              7.0
46     3029                     0.0             28.0
47     3395                    30.0             18.0
48     3904                     0.0             15.0
49     4006                     7.0              7.0
50     4006                     7.0              7.0
51     4006                     7.0              7.0
52     4006                     5.0              7.0
53     4006                     4.0              7.0
54     4122                     8.0              7.0
55     4141                    30.0             30.0
56     4343                     0.0             15.0
57     4553                    29.0             30.0
58     4845                    12.0             17.0
59     6866                     0.0             30.0

#### Frequency flag 

In [24]:
ords_prods_merge.loc[ords_prods_merge['order_frequency'] >= 20 , 'frequency_flag'] = 'Non-frequent customer'

In [25]:
ords_prods_merge.loc[(ords_prods_merge['order_frequency'] <= 20) & (ords_prods_merge['order_frequency'] > 10), 'frequency_flag'] = 'Regular customer'

In [26]:
ords_prods_merge.loc[ords_prods_merge['order_frequency'] <= 10 , 'frequency_flag'] = 'Frequent customer'

In [27]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: frequency_flag, dtype: int64

In [28]:
ords_prods_merge.shape

(32404859, 25)

In [29]:
#Export the newly created dataframe .
ords_prods_merge.to_pickle(os.path.join(path, 'chapter4','Prepared data', 'ords_prods_flag.pkl'))